### PR TITLE
Initial merge of typed protocols 

### PR DESCRIPTION
**Pull request by dcoutts**
This brings the first parts of the typed protocol framework into the current code base.

It adds the core typed protocol `Peer` type, plus the ping pong and chain sync protocols.

It also adds a view of the protocol without the strong typing on the protocol states. This is a slightly different way of factoring the channel functionality compare with the original branch. It is quite close to the existing `Pipe` demo. It keeps the basic form of a `Peer` but hides the strong types, meaning that there is one dynamic test on receive (rather than this being guaranteed by the type), but which is the right form for pushing in messages from the wire.

### CODE REVIEW

File: src/Ouroboros/Network/Protocol/ChainSync/Server.hs [hunk 1]
```diff
@@ -0,0 +1,115 @@
+{-# LANGUAGE GADTs           #-}
+{-# LANGUAGE DataKinds       #-}
+{-# LANGUAGE NamedFieldPuns  #-}
+
+-- | A view of the chain synchronisation protocol from the point of view of the
+-- server.
+--
+-- This provides a view that uses less complex types and should be easier to
+-- use than the underlying typed protocol itself.
```
Comment (coot) [inline src/Ouroboros/Network/Protocol/ChainSync/Server.hs]: This comments seems not right, I guess copy pasted from `Type.hs`.

**Comment by avieth** [comment]
Why not use #10 ?

**Comment by avieth** [comment]
> Why not use #10 ?

That one has the core typed-transitions definitions in a separate package, which is in my opinion a good thing. It's nothing to do with ouroboros in particular.

File: src/Ouroboros/Network/Protocol/Untyped.hs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  Expected   :: msg from to -> TransitionFrom msg from
+  Unexpected :: TransitionFrom msg from
+
+castTransition
+  :: forall st (msg :: st -> st -> Type) (from :: st) .
+     ( Typeable from )
+  => Proxy from
+  -> SomeMessage msg
+  -> TransitionFrom msg from
+castTransition _proxy (SomeMessage (m :: msg from' to)) =
```
Comment (avieth) [inline src/Ouroboros/Network/Protocol/Untyped.hs]: Naming inconsistency: `castTransition`, `TransitionFrom`, but `SomeMessage`. Are we calling them messages or transitions? I like "transition". "message" carries connotations from networking/IPC. There's nothing inherently IPC about the typed-transitions style. A typed-transitions presentation can be run in  a single process.

File: src/Ouroboros/Network/Protocol/ChainSync/Direct.hs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+-- | The 'ClientStream m' and 'ServerStream m' types are complementary. The
+-- former can be used to feed the latter directly, in the same thread.
+-- That's demonstrated here by constructing 'direct'.
+--
+direct :: Monad m
+       => ChainSyncServer header point m a
+       -> ChainSyncClient header point m b
+       -> m (a, b)
+
+--TODO: note that currently these protocols are non-terminating.
```
[the remark this answers is not in the document]
Comment (avieth) [reply src/Ouroboros/Network/Protocol/ChainSync/Direct.hs]: This is a very important point IMO. Both sides ought to be able to terminate. It makes testing so much more pleasant, and obviates the need for probes.

File: src/Ouroboros/Network/Protocol/Untyped.hs (hunk 2 above)
Comment (dcoutts) [reply src/Ouroboros/Network/Protocol/Untyped.hs]: Yep fair point. Will review names there.

I'm trying to use the name message, precisely becaue this is about message passing and I think it helps with comprehension. You're right of course that it's in principle more general, but that's not how we're using it here. We're using it for protocols (which usually but don't have to be between concurrent processes).

Indeed I plan to rename Hole -> Effect, and Yield/Await to Send/Recv. I relaise it pains you to see us specialise the naming to the protocol/IPC use case, but I think it's worth it to make it all more comprehensible. It was honestly a real slog at first (which is partly why it was so long before proper review). The naming has something to do with that.

File: src/Ouroboros/Network/Protocol/ChainSync/Server.hs (hunk 1 above)
Comment (dcoutts) [reply src/Ouroboros/Network/Protocol/ChainSync/Server.hs]: No this isn't a copy&paste'o. It's the same comment for the client and server side, but that's true in both cases.

File: src/Ouroboros/Network/Protocol/ChainSync/Direct.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (dcoutts) [reply src/Ouroboros/Network/Protocol/ChainSync/Direct.hs]: I agree we should extend it for just those reasons. I was starting with the protocol as it exists now, which does not have any in-band shutdown message.

I'd like to see if we can make certain shutdown messages be effectively channel shutdown. Then I'll be happier about doing it this way in the typed protocol state transitions.

File: src/Ouroboros/Network/Protocol/ChainSync/Type.hs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+data ChainSyncMessage header point from to where
+
+  -- | Request the next update from the producer. The response can be a roll
+  -- forward, a roll back or wait.
+  --
+  MsgRequestNext :: ChainSyncMessage header point StIdle (StNext StCanAwait)
+
+  -- | Acknowledge the request but require the consumer to wait for the next
+  -- update. This means that the consumer is synced with the producer, and
+  -- the producer its waiting for its own chain state to change.
```
[the remark this answers is not in the document]
Comment (dcoutts) [reply src/Ouroboros/Network/Protocol/ChainSync/Type.hs]: Well spotted, it should be `the producer is waiting`.

**Comment by dcoutts** [comment]
@coot how do we fix the nix build here?

Note that it builds fine with cabal because cabal has found a solution using QuickCheck-2.12 and psqueues-0.2.7, even though the testsuite for psqueues requires QuickCheck < 2.12. That's because cabal's solver knows that it's ok to use a different version of QuickCheck for our test suite vs the psqueus test suite, since the two are not linked together. It knows we don't need global consistency here. Can we do the same thing with nix?

**Comment by coot** [comment]
I am not sure if there is a good fix, I can have a hot fix that disables running tests on `psqueues`, and I think this will fix it.  I already had the same problem with `aeson`. We are using a version of `aeson` which I forked to use `QuickCheck-2.12` and I made a [PR](https://github.com/bos/aeson/pull/676), then it turned out the aeson CI and its mainteiner have a problem with build times on `GHC-8.6.1` which I could not reproduce.  I'll ask tomorrow our nix guys how to solve this kind of problems.

**Comment by coot** [comment]
@dcoutts  I fixed the issues here: #38 (it's building now on CI, but the previous version run successfully and I only fixed a documentation typo).  I can just apply the changes here and we can merge this branch (since it has all the conversation) or the other one, as you prefer (I didn't want to make a rebase by a surprise here, hence a separate PR )

**Comment by coot** [comment]
I think we can close this PR for now.  I think everything was merged into `avieth/typed-transitions` which is in `master` now.

**Pull request closed without merging**